pyogrio/_io.pyx

```diff
@@ -2178,3 +2180,182 @@ def ogr_write(
         exc = exc_check()
         if exc:
             raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+
+def ogr_write_arrow(
+    str path,
+    str layer,
+    str driver,
+    arrow_obj,
+    str crs,
+    str geometry_type,
+    str geometry_name,
+    str encoding,
+    object dataset_kwargs,
+    object layer_kwargs,
+    bint append=False,
+    dataset_metadata=None,
+    layer_metadata=None,
+):
+    IF CTE_GDAL_VERSION < (3, 8, 0):
+        raise RuntimeError("Need GDAL>=3.8 for Arrow write support")
+
+    cdef OGRDataSourceH ogr_dataset = NULL
+    cdef OGRLayerH ogr_layer = NULL
+
+    if not encoding:
+        encoding = locale.getpreferredencoding()
+
+    layer_created = create_ogr_dataset_layer(
+        path, layer, driver, crs, geometry_type, encoding,
+        dataset_kwargs, layer_kwargs, append,
+        dataset_metadata, layer_metadata,
+        &ogr_dataset, &ogr_layer,
+    )
+    if geometry_name:
+        opts = {"GEOMETRY_NAME": geometry_name}
+    else:
+        opts = {}
+    cdef char **options = dict_to_options(opts)
+
+    stream_capsule = arrow_obj.__arrow_c_stream__()
+    write_arrow_stream_capsule(ogr_layer, stream_capsule, geometry_name, options)
+
+    ### Final cleanup
+    if ogr_dataset != NULL:
+        GDALClose(ogr_dataset)
+
+        # GDAL will set an error if there was an error writing the data source
+        # on close
+        exc = exc_check()
+        if exc:
+            raise DataSourceError(f"Failed to write features to dataset {path}; {exc}")
+
+
+IF CTE_GDAL_VERSION >= (3, 8, 0):
```

Comment:
I think that's what @kylebarron originally tried (mimicking how we did it for reading), but didn't get that working and so ended up with the current approach of putting the whole function in the IF. But can try once more with the other approach

Reply:
> Do you anticipate `write_arrow_stream_capsule` and `create_fields_from_arrow_schema` being reusable? It would make `ogr_write_arrow` a bit longer, but in theory those could just be in the body of that function.

I inlined `write_arrow_stream_capsule`, but kept `create_fields_from_arrow_schema` separate (I don't think we will directly reuse it elsewhere, but that feels a bit easier to follow code-organization wise, because it is a nice separate piece of functionality)

> It looks like at least `is_geometry_field` and `get_extension_metadata` could just be regular top-level functions and not need to be in this block.

Yes, moved those out of the IF block (and also dedented `create_fields_from_arrow_schema` by adding an IF raising an error inside it)